File: python/tvm/contrib/pipeline_executor.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
"""Pipeline executor that executes a series of modules in a pipeline fashion."""
import json
import os
import time
from tvm import runtime
from tvm._ffi import get_global_func
from tvm.contrib import graph_executor


def pipeline_executor_enabled():
    """Check if the pipeline executor is enabled.

    Return
    -------
    enable: bool
        Return whether the pipeline executor is enabled.
    """
    return get_global_func("tvm.pipeline_executor.create", allow_missing=True) is not None


class PipelineModule(object):
    """Wrapper of runtime module, caller can use this module to set parameters and get outputs.

    Parameters
    ----------
    module : Union[PipelineExecutorFactoryModule, Module]
        Common interface for pipeline executor factory modules or Module.
    """

    def __init__(self, module):
        if isinstance(module, PipelineExecutorFactoryModule):
            self.module = module.get_pipeline_executor_module()
        else:
            self.module = module
        # Get the packed functions from the pipeline executor.
        self._get_params_group_pipeline_map = self.module["get_params_group_pipeline_map"]
        self._run = self.module["run"]
        self._set_param = self.module["set_param"]
        self._set_input = self.module["set_input"]
        self._get_input = self.module["get_input"]
        self._get_output = self.module["get_output"]
        self._get_num_outputs = self.module["get_num_outputs"]
        self._get_num_inputs = self.module["get_num_inputs"]
        self._get_input_pipeline_map = self.module["get_input_pipeline_map"]
        self._get_pipe_execute_count = self.module["get_execute_count"]

    def run(self):
        """Run the pipeline executor."""
        self._run()

    def get_input_pipeline_map(self, name):
        """Using the "name" to get the corresponding subgraph index and also get the "input name"
        of the corresponding subgraph interface.
        Returns
        -------
        input map: Array[str]
            Returning the index and "input name" of the subgraph.
        """
        return self._get_input_pipeline_map(name)

    def get_params_group_pipeline_map(self, name):
        """Use the name of the parameters group to get the corresponding runtime module index.

        Parameters
        ----------
        name: str
            The parameter group name.

        Returns
        -------
        module_index: int
            The index of the runtime module.
        """
        return self._get_params_group_pipeline_map(name)

    def set_input(self, key, value):
        """Set the input via input name.

        Parameters
        ----------
        key : str
            The input name
        value : array_like.
            The input value
        """
        self._set_input(key, value)

    def set_params(self, params_group_name, params_data):
        """Set the parameter group value given the parameter group name. Note that the parameter
        group name is declared in the pipeline executor config.

        Parameters
        ----------
        params_group_name : str
            The parameters group name.

        params_data : Dict[str, NDArray]
            A map from parameter name to data.
        """
        if not params_data:
            raise RuntimeError('"params_data is empty!"')

        for key, val in params_data.items():
            self._set_param(params_group_name, key, val)

    def get_input(self, key):
        """Get the input via an input name.
        Parameters
        ----------
        key : str
            The input key
        Returns
        -------
        data : NDArray
            The input data.
        """
        return self._get_input(key)

    def get_output(self, synchronize=True, sleep_interval=0.001):
        """Get the output.
        Returns
        -------
        data : Array[NDArray]
            A list of output data.
        synchronize : BOOL
            Whether to do a synchronize poll.
        sleep_interval : Float32
            When doing the synchronize loop poll, how many seconds the loop should sleep for yield.
        """
        outputs = []
        if not synchronize:
            outputs = self._get_output()
        else:
            while not outputs:
                outputs = self._get_output()
                time.sleep(sleep_interval)

        return outputs

    @property
    def num_executing_pipeline(self):
        """Getting the count of running pipeline.
        Returns
        -------
        count : int
            The count of running pipeline.
        """
        return self._get_pipe_execute_count()

    @property
    def num_outputs(self):
        """Get the number of outputs.
        Returns
        -------
        count : int
            The number of outputs.
        """
        return self._get_num_outputs()

    @property
    def num_inputs(self):
        """Get the number of inputs
        Returns
        -------
        count : int
            The number of inputs
        """
        return self._get_num_inputs()

    @staticmethod
    def load_library(config_file_name):
        """Import files to create a pipeline executor.

        Parameters
        ----------
        config_file_name : str
            Path and name of the configuration file, the configuration file contains the
            disk path of the parameter file, library file, and JSON file.
        """
        with open(config_file_name, "r") as file_handle:
            config = file_handle.read()
        config = json.loads(config)
        if "load_config" not in config or "pipeline_config" not in config:
            raise RuntimeError(
                f'"load_config" or "pipeline_config" is missing in {config_file_name}'
            )

        # The config file used to load library, prameters, and JSON files.
        with open(config["load_config"], "r") as file_handle:
            load_config = file_handle.read()

        # The config file used to load pipeline compute config.
        with open(config["pipeline_config"], "r") as file_handle:
            pipeline_config = file_handle.read()

        # Load a PipelineExecutor from the disk files.
        load_library = get_global_func("tvm.pipeline_executor.load", allow_missing=False)
        module = load_library(load_config, pipeline_config)

        return PipelineModule(module)


class PipelineExecutorFactoryModule(object):
    """Common interface for pipeline executor factory modules.

    Parameters
    ----------
    pipeline_mods : List[GraphExecutorFactoryModule]
        List of GraphExecutorFactoryModule.

    mod_config : Dict[int, Dict[str, Any]]
        Modules dependency configuration information.

    """

    def __init__(self, pipeline_mods, mods_config):
        self.pipeline_mods = pipeline_mods
        self.mods_config = mods_config
        self.module = None

    def get_pipeline_executor_module(self):
        """Get the pipeline executor module.

        Returns
        -------
        module : Module
            Common interface for pipeline executor factory Module.
        """
        if not self.module:
            graph_executors, config = self.graph_executor_create(
                self.pipeline_mods, self.mods_config
            )
            self.pipeline_create = get_global_func(
                "tvm.pipeline_executor.create", allow_missing=False
            )
            self.module = self.pipeline_create(graph_executors, config)
        return self.module

    def graph_executor_create(self, pipeline_mods, mod_config):
        """Create graph_executor list and return configuration as a json string.

        Parameters
        ----------
        pipeline_mods : List[GraphExecutorFactoryModule]
          List of GraphExecutorFactoryModule

        mod_config : Dict[str, Any]
            Modules dependency configuration information.

        Returns
        -------
        mods : List[Module]
            The Module list.

        mod_config : str
            The Modudle configuration.
        """
        # Should store modules in the list named 'mods' in index order.
        mods = [None for _ in range(len(pipeline_mods))]
        for lib_index in pipeline_mods:
            pipeline_lib = pipeline_mods[lib_index]["lib"]
            dev = pipeline_mods[lib_index]["dev"]
            lib = graph_executor.GraphModule(pipeline_lib["default"](dev))
            # Return a module list sorted by lib_index.
            mods[lib_index] = lib.module

        return mods, json.dumps(mod_config)

    def export_library(self, directory_path):
        """Export the pipeline executor into disk files.

        Parameters
        ----------
        directory_path : str
            Export the files to this directory.
        """
        if not self.pipeline_mods:
            raise RuntimeError("The pipeline executor has not been initialized.")

        # Check if the directory_path exists.
        if not os.path.exists(directory_path):
            raise RuntimeError("The directory {directory_path} does not exist.")
        # Create an load configuration.
        load_config_file_name = f"{directory_path}/load_config"
        pipeline_config_file_name = f"{directory_path}/pipeline_config"
        config = {}
        config["load_config"] = load_config_file_name
        config["pipeline_config"] = pipeline_config_file_name
        load_config = []
        # Export the library, JSON, and parameter into files, then export these files path
        # into a configuration file.
        for lib_index in self.pipeline_mods:
            mconfig = {}
            mconfig["mod_idx"] = lib_index
            mconfig["lib_name"] = f"{directory_path}/lib{lib_index}.so"
            mconfig["json_name"] = f"{directory_path}/json{lib_index}"
            mconfig["params_name"] = f"{directory_path}/params{lib_index}"
            mconfig["dev"] = (
                f"{self.pipeline_mods[lib_index]['dev'].device_type},"
                f"{self.pipeline_mods[lib_index]['dev'].device_id}"
            )
            # Get the graph, lib, and parameters from GraphExecutorFactoryModule.
            lib = self.pipeline_mods[lib_index]["lib"]
            # Export the lib, graph, and parameters to disk.
            if self.pipeline_mods[lib_index]["export_cc"]:
                lib.export_library(
                    mconfig["lib_name"], cc=self.pipeline_mods[lib_index]["export_cc"]
                )
            else:
                lib.export_library(mconfig["lib_name"])

            with open(mconfig["json_name"], "w") as file_handle:
                file_handle.write(lib.graph_json)
            with open(mconfig["params_name"], "wb") as file_handle:
                file_handle.write(runtime.save_param_dict(lib.params))

            load_config.append(mconfig)

        with open(load_config_file_name, "w") as file_handle:
            json.dump(load_config, file_handle)

        with open(pipeline_config_file_name, "w") as file_handle:
            json.dump(self.mods_config, file_handle)

        config_file_name = f"{directory_path}/config"
        with open(config_file_name, "w") as file_handle:
            json.dump(config, file_handle)

        return config_file_name

File: src/runtime/pipeline/pipeline_executor.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/*!
 * \file pipeline_executor.cc
 */
#include "pipeline_executor.h"
namespace tvm {
namespace runtime {
/*!
 * \brief Give frontends an access to packed functions.
 * \param name The name of the function.
 * \param sptr_to_self The pointer to the module node.
 * \return The corresponding packed function.
 */
PackedFunc PipelineExecutor::GetFunction(const String& name,
                                         const ObjectPtr<Object>& sptr_to_self) {
  if (name == "get_num_outputs") {
    return PackedFunc(
        [sptr_to_self, this](TVMArgs args, TVMRetValue* rv) { *rv = this->NumOutputs(); });
  } else if (name == "get_num_inputs") {
    return PackedFunc(
        [sptr_to_self, this](TVMArgs args, TVMRetValue* rv) { *rv = this->NumInputs(); });
  } else if (name == "get_input_pipeline_map") {
    return PackedFunc([sptr_to_self, this](TVMArgs args, TVMRetValue* rv) {
      if (String::CanConvertFrom(args[0])) {
        *rv = this->GetInputPipeplineMap(args[0].operator String());
      } else {
        LOG(FATAL) << "Function only support the input name value in the form of string";
      }
    });
  } else if (name == "get_params_group_pipeline_map") {
    return PackedFunc([sptr_to_self, this](TVMArgs args, TVMRetValue* rv) {
      if (String::CanConvertFrom(args[0])) {
        *rv = this->GetParamsGroupPipelineMap(args[0].operator String());
      } else {
        LOG(FATAL) << "Function only support the input name value in the form of string";
      }
    });
  } else if (name == "set_param") {
    return PackedFunc([sptr_to_self, this](TVMArgs args, TVMRetValue* rv) {
      if (String::CanConvertFrom(args[0]) && String::CanConvertFrom(args[1])) {
        this->SetParam(args[0].operator String(), args[1].operator String(), args[2]);
      } else {
        LOG(FATAL) << "Function only support the parameter name and the key in the form of string";
      }
    });
  } else if (name == "set_input") {
    return PackedFunc([sptr_to_self, this](TVMArgs args, TVMRetValue* rv) {
      if (String::CanConvertFrom(args[0])) {
        this->SetInput(args[0].operator String(), args[1]);
      } else {
        LOG(FATAL) << "Function only support the input name value in the form of string";
      }
    });
  } else if (name == "get_input") {
    return PackedFunc([sptr_to_self, this](TVMArgs args, TVMRetValue* rv) {
      if (String::CanConvertFrom(args[0])) {
        *rv = this->GetInput(args[0].operator String());
      } else {
        LOG(FATAL) << "Function only support the input name value in the form of string";
      }
    });
  } else if (name == "get_output") {
    return PackedFunc(
        [sptr_to_self, this](TVMArgs args, TVMRetValue* rv) { *rv = this->GetOutput(); });
  } else if (name == "run") {
    return PackedFunc([sptr_to_self, this](TVMArgs args, TVMRetValue* rv) { this->Run(); });
  } else if (name == "get_execute_count") {
    return PackedFunc(
        [sptr_to_self, this](TVMArgs args, TVMRetValue* rv) { *rv = this->GetExecutionCount(); });
  } else {
    LOG(FATAL) << "Unknown packed function: " << name;
  }
}
/*!
 * brief Returns number of global inputs.
 */
int PipelineExecutor::NumInputs(void) { return input_connection_config_.GetInputNum(); }
/*!
 * \brief set input to the runtime module.
 * \param input_name The input name.
 * \param data_in The input data.
 */
void PipelineExecutor::SetInput(std::string input_name, DLTensor* data_in) {
  global_runtime_->SetPipelineInput(input_name, data_in);
}
/*!
 * \brief get input from the runtime module.
 * \param input_name The input name.
 * \return Return the input data for a specific input name.
 */
NDArray PipelineExecutor::GetInput(std::string input_name) {
  std::pair<int, int> indexs = this->GetInputIndex(input_name);
  if (indexs.first < 0 || indexs.first >= static_cast<int>(runtimes_.size())) {
    LOG(FATAL) << "input name " << input_name << " not found.";
  }
  return runtimes_[indexs.first]->GetInput(indexs.second);
}
/*!
 * \brief Getting a module index via a input parameters group name.
 * \param name The parameters group name.
 * \return int The module index.
 */
int PipelineExecutor::GetParamModuleIndex(const std::string& name) {
  return param_connection_config_[name];
}
/*!
 * \brief Using the global input name to get the index, and also get the input interface name
   of corresponding subgraph from the input connection configuration.
 * \param The global input name.
 * \return Returning the index and the input interface name of corresponding subgraph.
 */
Array<String> PipelineExecutor::GetInputPipeplineMap(std::string input_name) {
  std::pair<int, std::string> map = input_connection_config_[input_name];
  return {std::to_string(map.first), map.second};
}

/*!
 * \brief Return the module index for the parameters group name.
 * \param name The parameters group name.
 * \return int The module index.
 */
int PipelineExecutor::GetParamsGroupPipelineMap(const std::string& name) {
  return param_connection_config_[name];
}

/*!\brief Run the pipeline executor.*/
void PipelineExecutor::Run() { pipeline_scheduler_.PipelineRun(runtimes_); }
/*!
 * \brief return A list of global output data.
 */
Array<NDArray> PipelineExecutor::GetOutput(void) { return pipeline_scheduler_.PipelineGetOutput(); }
/*!
 * \brief Use the mod_config information to create a graph runtime list.
 * \param mod_config The config information that generates by the export library function call.
 */
std::vector<Module> PipelineExecutor::CreateGraphModules(const ModuleConfig& mod_config) {
  const PackedFunc* graph_executor_create = Registry::Get("tvm.graph_executor.create");
  std::vector<Module> ret;
  ret.resize(mod_config.size());
  for (auto config : mod_config) {
    // Load library.
    auto lib = Module::LoadFromFile(config.second.lib_name.c_str());

    // Read json.
    std::ifstream ifJson(config.second.json_name.c_str());
    if (ifJson.fail()) {
      LOG(FATAL) << "json file not found: " << config.second.json_name;
    }
    const std::string json((std::istreambuf_iterator<char>(ifJson)),
                           std::istreambuf_iterator<char>());

    // Create a graph executor.
    std::istringstream istr(config.second.dev);
    std::string str;
    int device_type = 1, device_id = 0;
    while (getline(istr, str, ';')) {
      std::istringstream istr_dev(str);
      std::string str_temp;
      if (getline(istr_dev, str_temp)) {
        device_type = stoi(str_temp);
      }
      if (getline(istr_dev, str_temp)) {
        device_id = stoi(str_temp);
      }
    }
    Module graph_module = (*graph_executor_create)(json, lib, device_type, device_id);

    // Load parameters.
    TVMByteArray params_arr;
    const char* params_file_name = config.second.params_name.c_str();
    std::ifstream if_param(params_file_name);
    if (if_param.fail()) {
      LOG(FATAL) << "params file not found: " << params_file_name;
    }
    const std::string params((std::istreambuf_iterator<char>(if_param)),
                             std::istreambuf_iterator<char>());
    params_arr.data = params.c_str();
    params_arr.size = params.length();
    auto load_params = graph_module.GetFunction("load_params");
    load_params(params_arr);

    // Put a graph executor module into the vector.
    ret[config.first] = graph_module;
  }
  return ret;
}
/*!
 * \brief Set a parameter into a graph module.
 * \param param_group_name The parameters group name.
 * \param param_key_name The parameter key name.
 * \param data_in The parameter data.
 */
void PipelineExecutor::SetParam(std::string param_group_name, std::string param_key_name,
                                DLTensor* data_in) {
  // Get the module index via the parameters group name.
  int module_index = this->GetParamModuleIndex(param_group_name);
  ICHECK(module_index >= 0 && module_index < static_cast<int>(runtimes_.size()))
      << "Parameter group name " << param_group_name << " does not exist.";
  auto runtime = runtimes_[module_index];
  // Get the parameter index via the param key name
  int index = runtime->GetInputIndex(param_key_name);
  ICHECK(index >= 0) << "Parameter name " << param_key_name << " does not exist in module "
                     << module_index;
  runtime->SetInput(index, data_in);
}
/*!
 * \brief Return the input index and module index for a given input name.
 * \param name The input name.
 * \return std::pair<int, int> A pair of module index and the input index.
 */
std::pair<int, int> PipelineExecutor::GetInputIndex(const std::string& name) {
  std::pair<int, std::string> index = input_connection_config_[name];
  auto gruntime = runtimes_[index.first];
  return std::make_pair(index.first, gruntime->GetInputIndex(index.second));
}
/*!
 * \brief Getting the count of running pipeline.
 */
int PipelineExecutor::GetExecutionCount() { return runtimes_.back()->GetExecutionCount(); }
/*!
 * \brief Initialize the pipeline executor with a list of modules to be pipelined
 *  and config in JSON format.
 * \param modules The module list used for building the pipeline.
 * \param pipeline_json The configuration of modules dependencies.
 */
void PipelineExecutor::Init(const std::vector<Module>& modules, const std::string& pipeline_json) {
  ICHECK(!modules.empty()) << "The graph executor module list is empty.";
  // Use JSONReader to load pipeline configuration.
  std::istringstream is(pipeline_json);
  dmlc::JSONReader reader(&is);
  this->LoadConfig(&reader);
  ICHECK(!pipeline_config_.Empty()) << "The pipeline config information is empty.";
  num_outputs_ = pipeline_config_.GetGlobalOutputNum();
  // Initialize the pipeline function class used for pipeline thread pool management
  // and schedule etc. This function returns a list of runtime.
  global_runtime_ =
      pipeline_scheduler_.PipelineInit(modules, pipeline_config_, input_connection_config_);
  runtimes_ = global_runtime_->GetRuntimeList();
  return;
}

Module PipelineExecutorCreate(const Array<Module>& m, const std::string& pipeline_json) {
  ICHECK(!m.empty()) << "The module list is empty.";
  auto exec = make_object<PipelineExecutor>();
  std::vector<Module> graph_modules;
  for (auto mod : m) {
    graph_modules.push_back(mod);
  }
  exec->Init(graph_modules, pipeline_json);
  return Module(exec);
}

Module PipelineExecutorLoad(const std::string& load_json, const std::string& pipeline_json) {
  auto exec = make_object<PipelineExecutor>();
  std::istringstream is(load_json);
  dmlc::JSONReader reader(&is);
  ModuleConfig& mod_config = exec->LoadModuleConfig(&reader);
  ICHECK(!mod_config.empty()) << "The module config is empty.";
  std::vector<Module> modules = exec->CreateGraphModules(mod_config);
  exec->Init(modules, pipeline_json);
  return Module(exec);
}

TVM_REGISTER_GLOBAL("tvm.pipeline_executor.create").set_body([](TVMArgs args, TVMRetValue* rv) {
  *rv = PipelineExecutorCreate(args[0], args[1]);
});

TVM_REGISTER_GLOBAL("tvm.pipeline_executor.load").set_body([](TVMArgs args, TVMRetValue* rv) {
  *rv = PipelineExecutorLoad(args[0], args[1]);
});
}  // namespace runtime
}  // namespace tvm
